### PR TITLE
[SYCL] USM Buffer Location Properties

### PR DESCRIPTION
**Pull request by sherry-yuan**
Lower level runtime's usm allocation API now supports pass in of buffer location property [1] defined in OpenCL spec [2]

For this feature to be accessible to users, sycl runtime now passes in usm::buffer_location property defined in sycl spec [3] into the opencl runtime calls, the property is only passed if the extension name occurs in the platform extension string. If the lower level runtime does not support such property, then the property will yield no effect.

This is a temporary solution to allow user to specify which memory location the device usm allocation should be in. The full solution will require retuning an annotated_ptr that carries compile time properties for further optimization. The full solution spec is in [4]

[1] https://github.com/intel/fpga-runtime-for-opencl/pull/46/files
[2] https://github.com/KhronosGroup/OpenCL-Docs/blob/master/extensions/cl_intel_mem_alloc_buffer_location.asciidoc
[3] https://github.com/intel/llvm/pull/5665
[4] https://github.com/intel/llvm/pull/5656

### CODE REVIEW

**Comment by sherry-yuan** [comment]
SYCL / Linux / HIP AMDGPU LLVM Test Suite fails not due to this change, it falled for a no-functional change in https://github.com/intel/llvm/pull/5656/files#

**Comment by steffenlarsen** [comment]
Note that this does not directly adhere to the [cl_intel_mem_alloc_buffer_location](https://github.com/KhronosGroup/OpenCL-Docs/blob/master/extensions/cl_intel_mem_alloc_buffer_location.asciidoc) proposal as this implementation uses SYCL 2020 properties instead of the properties introduced in the [sycl_ext_oneapi_properties](https://github.com/sherry-yuan/llvm/blob/malloc-properties/sycl/doc/extensions/proposed/sycl_ext_oneapi_properties.asciidoc) proposal (in development).

@gmlueck - Should we add a variant of this for SYCL 2020 property lists to prevent it from being blocked by the development of the compile-time property lists?

**Comment by sherry-yuan** [comment]
> Note that this does not directly adhere to the [cl_intel_mem_alloc_buffer_location](https://github.com/KhronosGroup/OpenCL-Docs/blob/master/extensions/cl_intel_mem_alloc_buffer_location.asciidoc) proposal as this implementation uses SYCL 2020 properties instead of the properties introduced in the [sycl_ext_oneapi_properties](https://github.com/sherry-yuan/llvm/blob/malloc-properties/sycl/doc/extensions/proposed/sycl_ext_oneapi_properties.asciidoc) proposal (in development).

Yes this is more of a stopgap solution for passing properties into the runtime. so that memories can be allocated in the right place. Full solution that uses [sycl_ext_oneapi_properties](https://github.com/sherry-yuan/llvm/blob/malloc-properties/sycl/doc/extensions/proposed/sycl_ext_oneapi_properties.asciidoc) is here; https://github.com/intel/llvm/pull/5656

**Comment by sherry-yuan** [comment]
We will need this to be merged by ~Mar 26 for this to be integrated internally for next release.

File: sycl/test/extensions/usm/usm_alloc_utility.cpp
Comment (gmlueck) [inline sycl/test/extensions/usm/usm_alloc_utility.cpp:87]:
```suggestion
    array = malloc_device<int>(
        N, q,
```

I think you introduced a bug here when you changed the call to use the template form of `malloc_device()`.  In the template form, the first parameter is the count of *elements*, not bytes.  Also, there is no need to cast the return type because it is already a pointer to the element type.

File: sycl/include/CL/sycl/detail/property_helper.hpp
Comment (gmlueck) [inline sycl/include/CL/sycl/detail/property_helper.hpp:51]: Shouldn't the new enum be put at the end?

Comment (sherry-yuan) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: The same change was introduced in another PR as well https://github.com/intel/llvm/pull/5604/files#r818076116

I will change the order if there is no compelling reason from Maxim.

Comment (sherry-yuan) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: I found the reason, there is a comparison somewhere where all other enum of the same type cannot exceed this last enum

https://github.com/intel/llvm/blob/54ede407edeb93b7e9334d1e725f48bf981b2965/sycl/include/CL/sycl/detail/property_list_base.hpp#L94

Comment (gmlueck) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: This doesn't break ABI compatibility with existing object files which assume `PropWithDataKindSize` has the value `5`?

Comment (sherry-yuan) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: Hi @alexbatashev , I wonder why is this restriction below added in your [change](https://github.com/intel/llvm/commit/f7d073d1c853a10c93acb0376e6f8ca726aa4058)?

https://github.com/intel/llvm/blob/54ede407edeb93b7e9334d1e725f48bf981b2965/sycl/include/CL/sycl/detail/property_list_base.hpp#L94-L97

It seems to say that whenever a new property kind is introduced, `PropWithDataKindSize` will have to change, which potentially breaks ABI when some object assumes it is some older value. Can this restriction be removed, or does no other binary uses this enum?

Comment (steffenlarsen) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: I believe the intention of the check is to prevent other types from impersonating properties, i.e. I cannot go make
```c++
struct ThisIsNotActuallyAProperty : PropertyWithDataBase {
  static constexpr int getKind() {
    return PropWithDataKind::PropWithDataKindSize;
  }
};
```
and then expect it to work as a property.

Incrementing `PropWithDataKindSize` does not break ABI nor API as the only use of it is in a template function in the header. This template function requires the property type as a template parameter, which means for it to be instantiated it must know the property and as such should also know the property kind. As long as you don't increment existing property kinds you are guaranteed that, even if different instantiations know different values for `PropWithDataKindSize` it will only fail for valid cases.

Comment (sherry-yuan) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: Got it, thanks @steffenlarsen :)

Comment (sherry-yuan) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: Resolving this thread given there is no problem.

**Review by gmlueck — APPROVED** [review]
LGTM, but someone else should review.

**Comment by cperkinsintel** [comment]
>> Once the new property is ready, this change will introduce an override malloc function that uses the new property.

The new properties were merged at the beginning of this week.  For both spec conformance and to reduce churn, should we not use them instead?  
@steffenlarsen

**Comment by sherry-yuan** [comment]
> The new properties were merged at the beginning of this week. For both spec conformance and to reduce churn, should we not use them instead?

Having that property list alone is not sufficient, we will also need annotated pointer to be ready. See the full solution spec for details: https://github.com/intel/llvm/pull/5656

I don't think the full solution will be ready by Mar 27 given the bandwidth, so this is a workaround that uses existing API and only introduce a simple property in this spec: https://github.com/intel/llvm/pull/5665

We went with this solution because it falls more closely with the expected timeline, bandwidth.

**Comment by sherry-yuan** [comment]
There was no error that was caused by this change, it seems like some test need to be updated because the name space of `property` was being confused with the new property list.

**Comment by steffenlarsen** [comment]
As @sherry-yuan pointed out there is another extension for using the new `properties` for doing this exactly, but it requires a new return type. We could make this an experimental feature and change the return type, which would mean we would break experimental ABI when the new return type is introduced. By using `sycl::property_list` instead we can deprecate this temporary API and allow users some time to transition to the "final" API.

File: sycl/include/sycl/ext/intel/experimental/usm_properties.hpp
Comment (steffenlarsen) [inline sycl/include/sycl/ext/intel/experimental/usm_properties.hpp:37]:
```suggestion
} // __SYCL_INLINE_NAMESPACE(cl)

```

File: sycl/source/detail/usm/usm_impl.cpp
Comment (steffenlarsen) [inline sycl/source/detail/usm/usm_impl.cpp:135]:
```suggestion
      bool IsBufferLocSupported = Dev.is_accelerator();
```

File: sycl/source/detail/usm/usm_impl.cpp
Comment (steffenlarsen) [inline sycl/source/detail/usm/usm_impl.cpp:138]:
```suggestion
      if (IsBufferLocSupported &&
          PropList.has_property<cl::sycl::ext::intel::experimental::property::
                                    usm::buffer_location>()) {
```
Checking `IsBufferLocSupported` first is cheap and can save us the additional check to `has_property`.

**Review by steffenlarsen — COMMENTED** [review]
Some minor comments, otherwise LGTM. Failures do not seem related.

**Review by cperkinsintel — APPROVED** [review]
LGTM

**Comment by sherry-yuan** [comment]
@intel/llvm-gatekeepers This PR is ready to merge.

The desired commit message body is the same as PR description & title.

Thanks,
Sherry

**Comment by sherry-yuan** [comment]
Note on precommit test failures, some of the tests does not use explicit namespace reference which caused some confusion with the new property, the test are fixed in this PR: https://github.com/intel/llvm-test-suite/pull/904

**Comment by sherry-yuan** [comment]
@intel/llvm-gatekeepers the [tests](https://github.com/intel/llvm-test-suite/pull/904) are fixed, but I currently don't have permissions to re-run the workflows. The tests should pass.

**Comment by sherry-yuan** [comment]
Thank you all for helping to get this in 😊!

**Pull request merged**
